File: 141LinkedListCycleDetection/LinkedListCycleDetection.go
package main

import (
	"fmt"
	nodestructure "leetcode/node_structure"
)

func hasCycle1(head *nodestructure.ListNode) bool {
	cur := head
	s := map[*nodestructure.ListNode]struct{}{}
	for nil != cur {
		if _, ok := s[cur]; ok {
			return true
		}
		s[cur] = struct{}{}
		cur = cur.Next
	}
	return false
}

func hasCycle2(head *nodestructure.ListNode) bool {
	if nil == head {
		return false
	}
	slow := head
	fast := head
	slow = slow.Next
	fast = fast.Next
	if nil == fast {
		return false
	}
	fast = fast.Next
	for nil != slow && nil != fast && nil != fast.Next && slow != fast {
		slow = slow.Next
		fast = fast.Next.Next
	}
	return slow == fast
}

func main() {
	fmt.Println(
		hasCycle2(nodestructure.ListNodeFromList([]int{1, 2, 3, 4, 5})),
	)
}

File: 141LinkedListCycleDetection/LinkedListCycleDetection.py
# -*- coding: utf-8 -*-

import sys
import os
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))
from node_structure.ListNode import ListNode

class LinkedListCycleDetection:
    def hasCycle1(self, head: "ListNode") -> "bool":
        s = set()
        cur = head
        while cur:
            if cur in s:
                return True
            else:
                s.add(cur)
            cur = cur.next
        return False

    def hasCycle2(self, head: "ListNode") -> "bool":
        """
        two pointers
        """
        if not head: return False
        slow = fast = head
        slow = slow.next
        fast = fast.next
        if not fast: return False
        fast = fast.next
        while slow and fast and fast.next and slow != fast:
            slow = slow.next
            fast = fast.next.next
        return slow == fast


def main():
    test = LinkedListCycleDetection()
    print(test.hasCycle2(
        ListNode.fromList([1, 2, 5, 3, 2])
    ))
    print(test.hasCycle2(
        ListNode.fromList([1, 2, 3, 4, 5, 6, 7, 8])
    ))

if __name__ == "__main__":
    main()
